lib/libc/stdlib/malloc/mrs/mrs.c

```diff
@@ -296,6 +300,27 @@ static void *entire_shadow;
 static bool quarantining = true;
 static bool revoke_every_free = false;
 static bool revoke_async = false;
+static bool bound_pointers = false;
+
+/*
+ * Optionally apply strict bounds to pointers returned by malloc() and friends.
+ * This is technically UB since it means that the pointer passed to free() is
+ * not what we got from the allocator.  However, it may be useful for
+ * demonstrating CHERI's spatial safety guarantees in demos and so on.
+ *
+ * The default behaviour lets the allocator provide pointers with bounds
+ * covering the usable size of the allocation, beyond the requested size, which
+ * is friendlier to realloc() loops.  With revocation enabled, strict bounds
+ * would force a new allocation for each realloc() call, which can hurt quite a
+ * bit.
+ */
+static void *
+mrs_bound_pointer(void *p, size_t size)
+{
+	if (p != NULL && bound_pointers)
+		p = cheri_setbounds(p, size);
```

Comment:
I think we need to make sure that size is at least 1.

It's possible for an allocator to actually support zero sized allocations, but it needs a special pool so they can't be confused with an adjacent allocation manipulated to point to one past the end with bounds of 0. No current allocator does this (maybe out dlmalloc port does by accident).